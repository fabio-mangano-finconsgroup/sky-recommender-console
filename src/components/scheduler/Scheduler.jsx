import React from 'react';
import FullCalendar from '@fullcalendar/react';
import resourceTimelinePlugin from '@fullcalendar/resource-timeline';
import UpsertVodRec from '../../containers/upsert-vod-rec/UpsertVodRec';
import RecTooltipInfo from './rec-tooltip-info/RecTooltipInfo';
import useNotification from '../../hooks/useNotification';
import Spinner from '../spinner/Spinner';
import Modal from '../../components/modal/Modal';
import UpsertLinRec from '../../containers/upsert-lin-rec/UpsertLinRec';
import { Hidden } from './Scheduler.styled';
import { resources, recTypes } from './config';
import { mapForScheduler } from './Scheduler.helpers';
import { getRec } from '../../providers/rec-provider/RecProvider';
import './style.css';

const Scheduler = () => {
  // isloading is handled here in order to avoid infinite loop
  // https://github.com/fullcalendar/fullcalendar-react/issues/97
  const [recIsLoading, setRecIsLoading] = React.useState(false);
  const [openModal, setOpenModal] = React.useState(false);
  const [isEditing, setIsEditing] = React.useState(false);
  const [selectedRec, setSelectedRec] = React.useState(undefined);
  const [recType, setRecType] = React.useState(undefined);

  const { addAlert } = useNotification();

  const CalendarRef = React.useRef();

  const modalTitle = `${isEditing ? `EDIT ${recType}` : `NEW ${recType}`}`;

  const handleOpenModal = () => setOpenModal(true);

  const handleCloseModal = () => {
    setOpenModal(false);
    if (isEditing) {
      setIsEditing(false);
      setSelectedRec(false);
    }
  };

  const loadRec = React.useCallback(
    async (info, success, error) => {
      try {
        const res = await getRec({
          validFrom_gte: info.startStr,
          validTo_lte: info.endStr,
        });
        const data = mapForScheduler(res);
        success(data);
      } catch (err) {
        addAlert({
          text: 'An error occured during the loading of the scheduled recommendations.',
          title: 'Recommendations loading failed',
          type: 'error',
          id: Date.now(),
        });
        error(err);
      }
    },
    [addAlert],
  );

  const handleRecLoading = (isLoading) => setRecIsLoading(isLoading);

  const handleRecEdit = (eventInfo) => {
    setIsEditing(true);
    setRecType(eventInfo.event.extendedProps.extraProps.type);
    setSelectedRec(eventInfo.event);
    handleOpenModal();
  };

  const handleRecCreate = (type) => {
    setRecType(type);
    handleOpenModal();
  };

  const handleCRUDSuccess = () => {
    handleCloseModal();
    CalendarRef.current.getApi().refetchEvents();
  };

  const renderRecContent = (arg) => {
    return (
      <RecTooltipInfo
        data-test-tooltip-id={arg.event.extendedProps.extraProps.id}
        title={arg.event.title}
        recommendation={arg.event.extendedProps.extraProps.recommendation}
        startDateTime={arg.event.start}
        endDateTime={arg.event.end}
      />
    );
  };

  const renderEvent = (data) => {
    data.el.setAttribute(
      'data-testId',
      'eventId-' + data.event.extendedProps.extraProps.id,
    );
  };

  return (
    <>
      {recIsLoading && <Spinner height="65vh" />}
      <Hidden isLoading={recIsLoading}>
        <FullCalendar
          ref={CalendarRef}
          schedulerLicenseKey="CC-Attribution-NonCommercial-NoDerivatives"
          plugins={[resourceTimelinePlugin]}
          initialView="resourceTimelineMonth"
          height="700px"
          resourceAreaWidth="180px"
          resourceAreaHeaderContent="Clusters"
          resources={resources}
          headerToolbar={{
            left: `prev,next,today,resourceTimelineDay,resourceTimelineWeek,resourceTimelineMonth`,
            center: 'title',
            right: 'newVod newLin',
          }}
          customButtons={{
            newVod: {
              text: 'NEW VOD',
              click: () => handleRecCreate(recTypes.vod),
            },
            newLin: {
              text: 'NEW LIN',
              click: () => handleRecCreate(recTypes.lin),
            },
          }}
          nowIndicator
          eventContent={renderRecContent}
          expandRows
          eventDidMount={renderEvent}
          events={loadRec}
          eventClick={handleRecEdit}
          loading={handleRecLoading}
        />
        <Modal
          title={modalTitle}
          open={openModal}
          handleClose={handleCloseModal}
        >
          {!isEditing && recType === recTypes.vod && (
            <UpsertVodRec onSuccess={handleCRUDSuccess} />
          )}
          {!isEditing && recType === recTypes.lin && (
            <UpsertLinRec onSuccess={handleCRUDSuccess} />
          )}
          {isEditing && recType === recTypes.vod && selectedRec?.id && (
            <UpsertVodRec id={selectedRec?.id} onSuccess={handleCRUDSuccess} />
          )}
          {isEditing && recType === recTypes.lin && selectedRec?.id && (
            <UpsertLinRec id={selectedRec?.id} onSuccess={handleCRUDSuccess} />
          )}
        </Modal>
      </Hidden>
    </>
  );
};

export default Scheduler;
