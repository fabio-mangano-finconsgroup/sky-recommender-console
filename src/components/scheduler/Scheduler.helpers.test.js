import { mapForScheduler } from './Scheduler.helpers';
import recommendation from '../../../fixtures/recomendation-mock.json';

describe('Scheduler helpers', () => {
  it('should transform the given recommendations in the shape accepted by the scheduler', () => {
    expect(mapForScheduler(recommendation)).toMatchSnapshot();
  });

  it('should throw an exception if no recommendations are provided', () => {
    expect(() => mapForScheduler(undefined)).toThrowError(
      'No recommendations provided.',
    );
  });
});
