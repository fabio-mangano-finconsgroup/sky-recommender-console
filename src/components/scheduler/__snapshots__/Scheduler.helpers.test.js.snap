// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scheduler helpers should transform the given recommendations in the shape accepted by the scheduler 1`] = `
Array [
  Object {
    "end": "",
    "extraProps": Object {
      "cluster": "CL_CIN",
      "id": "4",
      "recommendation": Array [
        Object {
          "channel": "Sky Cinema HD",
          "id": "61GA",
          "linearSummaryLong": "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.",
          "position": "1",
          "resolution": "HD",
          "startProgram": "2021-12-26T15:00:00Z",
          "thumbnailType": "",
          "title": "Avatar",
          "type": "LIN",
          "verticalImageUrl": "",
        },
        Object {
          "channel": "Sky Cinema HD",
          "id": "61GK",
          "linearSummaryLong": "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.",
          "position": "1",
          "resolution": "SD",
          "startProgram": "2021-12-26T15:00:00Z",
          "thumbnailType": "",
          "title": "L'immortale",
          "type": "LIN",
          "verticalImageUrl": "",
        },
        Object {
          "channel": "Sky Cinema HD",
          "id": "61GP",
          "linearSummaryLong": "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.",
          "position": "2",
          "resolution": "HD",
          "startProgram": "2021-12-26T15:00:00Z",
          "thumbnailType": "",
          "title": "Avengers",
          "type": "LIN",
          "verticalImageUrl": "",
        },
        Object {
          "channel": "Sky Cinema HD",
          "id": "61GL",
          "linearSummaryLong": "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.",
          "position": "2",
          "resolution": "SD",
          "startProgram": "2021-12-26T15:00:00Z",
          "thumbnailType": "",
          "title": "Benvenuti al sud",
          "type": "LIN",
          "verticalImageUrl": "",
        },
      ],
      "type": "LIN",
      "validFrom": "",
      "validTo": "",
    },
    "id": "4",
    "resourceId": "lin-1",
    "start": "",
    "title": "LIN",
  },
  Object {
    "end": "",
    "extraProps": Object {
      "cluster": "CL_CIN",
      "id": "50",
      "recommendation": Array [
        Object {
          "channel": "Sky Cinema HD",
          "id": "61GA",
          "linearSummaryLong": "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.",
          "position": "1",
          "resolution": "HD",
          "startProgram": "2021-12-26T15:00:00Z",
          "thumbnailType": "",
          "title": "Avatar",
          "type": "LIN",
          "verticalImageUrl": "",
        },
        Object {
          "channel": "Sky Cinema HD",
          "id": "61GK",
          "linearSummaryLong": "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.",
          "position": "1",
          "resolution": "SD",
          "startProgram": "2021-12-26T15:00:00Z",
          "thumbnailType": "",
          "title": "L'immortale",
          "type": "LIN",
          "verticalImageUrl": "",
        },
        Object {
          "channel": "Sky Cinema HD",
          "id": "61GP",
          "linearSummaryLong": "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.",
          "position": "2",
          "resolution": "HD",
          "startProgram": "2021-12-26T15:00:00Z",
          "thumbnailType": "",
          "title": "Avengers",
          "type": "LIN",
          "verticalImageUrl": "",
        },
        Object {
          "channel": "Sky Cinema HD",
          "id": "61GL",
          "linearSummaryLong": "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.",
          "position": "2",
          "resolution": "SD",
          "startProgram": "2021-12-26T15:00:00Z",
          "thumbnailType": "",
          "title": "Benvenuti al sud",
          "type": "LIN",
          "verticalImageUrl": "",
        },
      ],
      "type": "LIN",
      "validFrom": "",
      "validTo": "",
    },
    "id": "50",
    "resourceId": "lin-1",
    "start": "",
    "title": "LIN",
  },
  Object {
    "end": "",
    "extraProps": Object {
      "cluster": "CL_NOT_CIN",
      "id": "35",
      "recommendation": Array [
        Object {
          "branding": "Sky Cinema HD",
          "endProgram": "2022-12-14T15:00:00Z",
          "id": "61GP6",
          "position": "1",
          "startProgram": "2022-01-24T12:00:00Z",
          "thumbnailType": "",
          "title": "L'isola che non c'è",
          "type": "VOD",
          "verticalImageUrl": "",
          "vodSummaryLong": "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.",
        },
        Object {
          "branding": "Sky Cinema HD",
          "endProgram": "2022-12-14T15:00:00Z",
          "id": "61GP6",
          "position": "2",
          "startProgram": "2022-01-24T12:00:00Z",
          "thumbnailType": "",
          "title": "L'isola che non c'è",
          "type": "VOD",
          "verticalImageUrl": "",
          "vodSummaryLong": "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.",
        },
        Object {
          "branding": "Sky Cinema HD",
          "endProgram": "2022-12-14T15:00:00Z",
          "id": "61GP6",
          "position": "3",
          "startProgram": "2022-01-24T12:00:00Z",
          "thumbnailType": "",
          "title": "L'isola che non c'è",
          "type": "VOD",
          "verticalImageUrl": "",
          "vodSummaryLong": "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.",
        },
        Object {
          "branding": "Sky Cinema HD",
          "endProgram": "2022-12-14T15:00:00Z",
          "id": "61GP6",
          "position": "4",
          "startProgram": "2022-01-24T12:00:00Z",
          "thumbnailType": "",
          "title": "L'isola che non c'è",
          "type": "VOD",
          "verticalImageUrl": "",
          "vodSummaryLong": "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.",
        },
        Object {
          "branding": "Sky Cinema HD",
          "endProgram": "2022-12-14T15:00:00Z",
          "id": "61GP6",
          "position": "5",
          "startProgram": "2022-01-24T12:00:00Z",
          "thumbnailType": "",
          "title": "L'isola che non c'è",
          "type": "VOD",
          "verticalImageUrl": "",
          "vodSummaryLong": "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.",
        },
      ],
      "type": "VOD",
      "validFrom": "",
      "validTo": "",
    },
    "id": "35",
    "resourceId": "vod-2",
    "start": "",
    "title": "VOD",
  },
  Object {
    "end": "",
    "extraProps": Object {
      "cluster": "CL_NOT_CIN",
      "id": "166",
      "recommendation": Array [
        Object {
          "branding": "Sky Cinema HD",
          "endProgram": "2022-12-14T15:00:00Z",
          "id": "61GP6",
          "position": "1",
          "startProgram": "2022-01-24T12:00:00Z",
          "thumbnailType": "",
          "title": "L'isola che non c'è",
          "type": "VOD",
          "verticalImageUrl": "",
          "vodSummaryLong": "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.",
        },
        Object {
          "branding": "Sky Cinema HD",
          "endProgram": "2022-12-14T15:00:00Z",
          "id": "61GP6",
          "position": "2",
          "startProgram": "2022-01-24T12:00:00Z",
          "thumbnailType": "",
          "title": "L'isola che non c'è",
          "type": "VOD",
          "verticalImageUrl": "",
          "vodSummaryLong": "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.",
        },
        Object {
          "branding": "Sky Cinema HD",
          "endProgram": "2022-12-14T15:00:00Z",
          "id": "61GP6",
          "position": "3",
          "startProgram": "2022-01-24T12:00:00Z",
          "thumbnailType": "",
          "title": "L'isola che non c'è",
          "type": "VOD",
          "verticalImageUrl": "",
          "vodSummaryLong": "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.",
        },
        Object {
          "branding": "Sky Cinema HD",
          "endProgram": "2022-12-14T15:00:00Z",
          "id": "61GP6",
          "position": "4",
          "startProgram": "2022-01-24T12:00:00Z",
          "thumbnailType": "",
          "title": "L'isola che non c'è",
          "type": "VOD",
          "verticalImageUrl": "",
          "vodSummaryLong": "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.",
        },
        Object {
          "branding": "Sky Cinema HD",
          "endProgram": "2022-12-14T15:00:00Z",
          "id": "61GP6",
          "position": "5",
          "startProgram": "2022-01-24T12:00:00Z",
          "thumbnailType": "",
          "title": "L'isola che non c'è",
          "type": "VOD",
          "verticalImageUrl": "",
          "vodSummaryLong": "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.",
        },
      ],
      "type": "VOD",
      "validFrom": "",
      "validTo": "",
    },
    "id": "166",
    "resourceId": "vod-2",
    "start": "",
    "title": "VOD",
  },
]
`;
